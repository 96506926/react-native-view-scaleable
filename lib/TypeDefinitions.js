// @flow

export type AnimatedWatchValue<T> = {
	value: T
};

export type LayoutEvent = {
	nativeEvent: {
		target: number,
		layout: {
			width: number,
			height: number,
			x: number,
			y: number
		}
	}
};

export type Touch = {
	pageX: number,
	pageY: number,
	locationX: number,
	locationY: number
};

export type TransformTranslateLayout = [
	{ translateX: AnimatedValue<number> },
	{ translateY: AnimatedValue<number> }
];

export type Event = {
	nativeEvent: {
		changedTouches: number,
		identifier: number,
		locationX: number,
		locationY: number,
		pageX: number,
		pageY: number,
		target: number,
		timestamp: number,
		touches: Array<Touch>
	}
};
export type GestureState = {
	numberActiveTouches: number,
	stateID: number,
	moveX: number,
	moveY: number,
	x0: number,
	y0: number,
	dx: number,
	dy: number,
	vx: number,
	vy: number
};

export type PanResponderDef = {
	panHandlers: {
		onStartShouldSetPanResponder: ?(e: Event, g: GestureState) => boolean,
		onMoveShouldSetPanResponder: ?(e: Event, g: GestureState) => boolean,
		onPanResponderGrant: (e: Event, g: GestureState) => void,
		onPanResponderMove: (e: Event, g: GestureState) => void,
		onPanResponderRelease: (e: Event, g: GestureState) => void,
		onPanResponderTerminate: (e: Event, g: GestureState) => void,
	}
};

type InterpolateOptions<I, O> = {
	inputRange: Array<I>,
	outputRange: Array<O>,
	extrapolate?: string
};

export type AnimatedValue<T> = {
	__value: number,
	setValue: (value: T) => void,
	setOffset: (value: T) => void,
	flattenOffset: () => void,
	addListener: (cb: (value: AnimatedWatchValue<T>) => void) => number,
	removeListener: (id: number) => void,
	removeAllListeners: () => void,
	interpolate: (obj: InterpolateOptions<T, any>) => AnimatedValue<any>
};
export type AnimatedValueXY = {
	x: AnimatedValue<number>,
	y: AnimatedValue<number>,
	setValue: (values: { x: number, y: number }) => void,
	getLayout: () => { top: number, left: number },
	setOffset: (values: { x: number, y: number }) => void,
	flattenOffset: () => void,
	removeAllListeners: () => void,
	getTranslateTransform: () => TransformTranslateLayout
};
