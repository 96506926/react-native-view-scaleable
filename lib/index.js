// @flow

export { default as ViewEditor } from './ViewEditor';
