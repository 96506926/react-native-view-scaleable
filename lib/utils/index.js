// @flow

export {
	centerTouches,
	pow2abs,
	distanceBetweenTouches,
	toDeg,
	angle,
	degreeToNumber,
	numberToDegree
} from './geometry';
