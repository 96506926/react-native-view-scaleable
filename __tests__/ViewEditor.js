// @flow
import 'react-native';
import React from 'react';
import { View, Text } from 'react-native';
import renderer from 'react-test-renderer';
import { ViewEditor } from '../lib';
import util from 'util';

function getValueForKey(arr, key) {
	return (arr.find((transform) => transform.hasOwnProperty(key)) || {})[key];
}

describe('ViewEditor Component', () => {
	let component = null;
	const onMoveCallback = jest.fn();

	beforeEach(() => {
		component = renderer.create(
			<ViewEditor
				minScale={1}
				maxScale={20}
				onMove={onMoveCallback}
			>
				<View>
					<Text>Hello World</Text>
				</View>
			</ViewEditor>
		);
	});
	it('should render tree', () => {
		// $FlowFixMe
		const componentJSON = component.toJSON();
		expect(componentJSON).toMatchSnapshot();
	});
	it('should contain transform prop', () => {
		// $FlowFixMe
		const componentJSON = component.toJSON();
		expect(componentJSON.props.style.transform).toBeInstanceOf(Array);
	});
	it('should add translateX and translateY props correctly', () => {
		// $FlowFixMe
		const componentJSON = component.toJSON();
		const translateX = getValueForKey(componentJSON.props.style.transform, 'translateX');
		const translateY = getValueForKey(componentJSON.props.style.transform, 'translateY');

		expect(translateX).toBeDefined();
		expect(translateY).toBeDefined();

		expect(translateX).toBe(0);
		expect(translateY).toBe(0);
	});
	it('should add scale and rotate angle correctly', () => {
		// $FlowFixMe
		const componentJSON = component.toJSON();
		const rotate = getValueForKey(componentJSON.props.style.transform, 'rotate');
		const scale = getValueForKey(componentJSON.props.style.transform, 'scale');

		expect(rotate).toBeDefined();
		expect(scale).toBeDefined();

		// expect(rotate).toBe('0deg');
		expect(scale).toBe(1);
	});
	it('should call _getTransforms function and return transform values', () => {
		// $FlowFixMe
		const instance = component.getInstance();
		expect(instance._getTransforms()).toBeInstanceOf(Object);
		expect(instance._getTransforms().transform.length).toBe(4);
	});
	it('should call onMove prop on panning', () => {
		// $FlowFixMe
		const instance = component.getInstance();
		const fakeEvent = {
			nativeEvent: {
				changedTouches: 0,
				identifier: 0,
				locationX: 0,
				locationY: 0,
				pageX: 0,
				pageY: 0,
				target: 0,
				timestamp: 0,
				touches: {
					pageX: 1,
					pageY: 1,
					locationX: 1,
					locationY: 1
				}
			}
		};
		const fakeGestureState = {
			numberActiveTouches: 1,
			stateID: 1,
			moveX: 1,
			moveY: 1,
			x0: 1,
			y0: 1,
			dx: 1,
			dy: 1,
			vx: 1,
			vy: 1
		};
		instance._handlePanResponderMove(fakeEvent, fakeGestureState);
		expect(onMoveCallback).toHaveBeenCalled();
	});
});
