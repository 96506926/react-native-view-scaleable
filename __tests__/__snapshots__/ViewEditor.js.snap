// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewEditor Component should render tree 1`] = `
<View
  onLayout={[Function]}
  onMoveShouldSetResponder={[Function]}
  onMoveShouldSetResponderCapture={[Function]}
  onResponderEnd={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderReject={[Function]}
  onResponderRelease={[Function]}
  onResponderStart={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  onStartShouldSetResponderCapture={[Function]}
  style={
    Object {
      "transform": Array [
        Object {
          "translateX": 0,
        },
        Object {
          "translateY": 0,
        },
        Object {
          "scale": 1,
        },
        Object {
          "rotate": "0deg0",
        },
      ],
    }
  }
>
  <View>
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
    >
      Hello World
    </Text>
  </View>
</View>
`;
